samples/extensions/khr/externalmemory/vulkan_utils.h

```diff
@@ -0,0 +1,382 @@
+                 platforms);
+
+    size_t cl_device_count = 0;
+    const char* uuid_khronos_extension[] = {
+        CL_KHR_DEVICE_UUID_EXTENSION_NAME
+    };
+    for (cl_uint cl_platform_id = 0; cl_platform_id < cl_platform_count;
+         ++cl_platform_id)
+    {
+        cl_uint cl_platform_devices_count = 0;
+        OCLERROR_RET(clGetDeviceIDs(platforms[cl_platform_id],
+                                    CL_DEVICE_TYPE_ALL, 0, NULL,
+                                    &cl_platform_devices_count),
+                     error, platforms);
+        for (cl_uint device_id = 0; device_id < cl_platform_devices_count;
+             ++device_id)
+        {
+            cl_device_id device;
+            OCLERROR_PAR(device = cl_util_get_device(
+                cl_platform_id, device_id, CL_DEVICE_TYPE_ALL, &error), error, platforms);
+            cl_device_count +=
+                check_khronos_extensions(device, uuid_khronos_extension, 1);
+        }
+    }
+
+    // For each OpenCL device, query its Vulkan-compatible device UUID and
+    // add it to the list of candidates. The device must support the
+    // cl_khr_device_uuid extension for us to be able to query the device's
+    // UUID.
+    struct cl_device_candidate* cl_candidates =
+        (struct cl_device_candidate*)malloc(
+            cl_device_count * sizeof(struct cl_device_candidate));
+    cl_device_count = 0;
+    for (cl_uint cl_platform_id = 0; cl_platform_id < cl_platform_count;
+         ++cl_platform_id)
+    {
+        cl_uint cl_platform_devices_count = 0;
+        OCLERROR_RET(clGetDeviceIDs(platforms[cl_platform_id],
+                                    CL_DEVICE_TYPE_ALL, 0, NULL,
+                                    &cl_platform_devices_count),
+                     error, candidates);
+
+        for (cl_uint cl_candidate_id = 0;
+             cl_candidate_id < cl_platform_devices_count;
+             ++cl_candidate_id, ++cl_device_count)
+        {
+            cl_device_id device = cl_util_get_device(
+                cl_platform_id, cl_candidate_id, CL_DEVICE_TYPE_ALL, &error);
+            if (check_khronos_extensions(device, uuid_khronos_extension, 1))
+            {
+                cl_uchar vk_candidate_uuid[CL_UUID_SIZE_KHR];
+                OCLERROR_RET(clGetDeviceInfo(device, CL_DEVICE_UUID_KHR,
+                                             CL_UUID_SIZE_KHR,
+                                             &vk_candidate_uuid, NULL),
+                             error, candidates);
+
+                struct cl_device_candidate candidate;
+                candidate.device = device;
+                memcpy(candidate.uuid, &vk_candidate_uuid,
+                       sizeof(cl_uchar) * CL_UUID_SIZE_KHR);
+                cl_candidates[cl_device_count] = candidate;
+            }
+        }
+    }
+
+    // Query the Vulkan physical devices available.
+    uint32_t vk_device_count;
+    VK_CHECK(vkEnumeratePhysicalDevices(instance, &vk_device_count, NULL));
```

Comment:
FWIW the C version of the sample crashes here on my Linux system.  I haven't debugged further.

```gdb
Thread 1 "externalmemory" received signal SIGSEGV, Segmentation fault.
0x00007ffff7ca171f in unlink_chunk (p=p@entry=0x555555723dc0, av=0x7ffff7e19c80 <main_arena>) at ./malloc/malloc.c:1628
1628	./malloc/malloc.c: No such file or directory.
(gdb) bt
#0  0x00007ffff7ca171f in unlink_chunk (p=p@entry=0x555555723dc0, av=0x7ffff7e19c80 <main_arena>) at ./malloc/malloc.c:1628
#1  0x00007ffff7ca462b in _int_malloc (av=av@entry=0x7ffff7e19c80 <main_arena>, bytes=bytes@entry=64) at ./malloc/malloc.c:4307
#2  0x00007ffff7ca5139 in __GI___libc_malloc (bytes=64) at ./malloc/malloc.c:3329
#3  0x00007fffe75b7f65 in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#4  0x00007fffe75b812b in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#5  0x00007fffe75b9643 in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#6  0x00007fffe74fe397 in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#7  0x00007fffe74fe982 in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#8  0x00007fffe748eea9 in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#9  0x00007fffe729cd3c in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#10 0x00007fffe747c3c1 in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#11 0x00007fffe747cef1 in ?? () from /usr/lib/x86_64-linux-gnu/libvulkan_intel.so
#12 0x00007ffff7f49718 in ?? () from /lib/x86_64-linux-gnu/libvulkan.so.1
#13 0x00007ffff7f49dec in ?? () from /lib/x86_64-linux-gnu/libvulkan.so.1
#14 0x00007fffed99b40c in ?? () from /lib/x86_64-linux-gnu/libVkLayer_MESA_device_select.so
#15 0x00007ffff7f53e41 in vkEnumeratePhysicalDevices () from /lib/x86_64-linux-gnu/libvulkan.so.1
#16 0x000055555555736d in find_suitable_device (instance=0x5555555623d0, required_device_extensions=0x555555561050 <required_device_extensions>, required_device_extensions_count=2)
    at /home/bashbaug/git/opencl-sdk/samples/extensions/khr/externalmemory/vulkan_utils.h:301
#17 0x0000555555557d76 in main (argc=1, argv=0x7fffffffe008) at /home/bashbaug/git/opencl-sdk/samples/extensions/khr/externalmemory/main.c:264
(gdb) 
```

Reply:
Hmm for the Linux environments tested it works fine ([this](https://github.com/StreamHPC/OpenCL-SDK/blob/external-mem-sample/.github/workflows/presubmit.yml#L31-L126) is the matrix of combinations tested), could you provide more info about your setup to try to reproduce this?

Reply:
I'm still getting this crash on my system with the latest code.

> could you provide more info about your setup to try to reproduce this?

It's a pretty standard Ubuntu 22.04 system, nothing too fancy.

The main thing that is a little different is that it has several GPUs in it, including two discrete GPUs and one integrated GPU.  I can't think of a reason why that might cause this failure, though, and only with the C version of the sample?  The C++ version is working fine.

Reply:
I think I found the problem.

The issue is that one of my OpenCL devices (POCL) does not support the `cl_khr_device_uuid` extension, so it does not increment `cl_device_count` when determining the number of candidate devices (the initial loop in `find_suitable_device`).  When building the list of candidates (the second loop in `find_suitable_device`) the `cl_device_count` gets incremented unconditionally, though.  This means instead of skipping the POCL device entirely, there is a "hole" in the candidate device list, and we end up writing more candidate devices than we've allocated.

The fix is either to increment the `cl_device_count` conditionally when building the candidate device list, consistent with the way the number of candidates was computed.

Reply:
Certainly, the logic was faulty as you describe. I added a commit fixing this, so that:
1. We do not continue checking the Vulkan-compatible devices if there were no OpenCL-compatible devices to begin with (`cl_device_count == 0`)
2. While checking the Vulkan-compatible devices we only increment the `cl_device_count` if the device supports the required extension